src/Compilers/Core/CodeAnalysisTest/AnalyzerConfigTests.cs

```diff
@@ -664,6 +657,194 @@ public void BadNumberRanges()
+            Assert.True(matcher.IsMatch("/abc.fs"));
+            Assert.True(matcher.IsMatch("/abc.^s"));
+            Assert.False(matcher.IsMatch("/abc.vs"));
+            Assert.False(matcher.IsMatch("/abc.xs"));
+            Assert.False(matcher.IsMatch("/abc.vxs"));
+        }
+
+        [Fact]
+        public void CharacterClassRange()
+        {
+            var matcher = TryCreateSectionNameMatcher("[0-9]x").Value;
+            Assert.Equal("^.*/[0-9]x$", matcher.Regex.ToString());
+
+            Assert.True(matcher.IsMatch("/0x"));
+            Assert.True(matcher.IsMatch("/1x"));
+            Assert.True(matcher.IsMatch("/9x"));
+            Assert.False(matcher.IsMatch("/yx"));
+            Assert.False(matcher.IsMatch("/00x"));
+        }
+
+        [Fact]
+        public void CharacterClassNegativeRange()
+        {
+            var matcher = TryCreateSectionNameMatcher("[!0-9]x").Value;
+            Assert.Equal("^.*/[^0-9]x$", matcher.Regex.ToString());
+
+            Assert.False(matcher.IsMatch("/0x"));
+            Assert.False(matcher.IsMatch("/1x"));
+            Assert.False(matcher.IsMatch("/9x"));
+            Assert.True(matcher.IsMatch("/yx"));
+            Assert.False(matcher.IsMatch("/00x"));
+        }
+
+        [Fact]
+        public void CharacterClassRangeAndChoice()
+        {
+            var matcher = TryCreateSectionNameMatcher("[ab0-9]x").Value;
+            Assert.Equal("^.*/[ab0-9]x$", matcher.Regex.ToString());
+
+            Assert.True(matcher.IsMatch("/ax"));
+            Assert.True(matcher.IsMatch("/bx"));
+            Assert.True(matcher.IsMatch("/0x"));
+            Assert.True(matcher.IsMatch("/1x"));
+            Assert.True(matcher.IsMatch("/9x"));
+            Assert.False(matcher.IsMatch("/yx"));
+            Assert.False(matcher.IsMatch("/0ax"));
+        }
+
+        [Fact]
+        public void CharacterClassOpenEnded()
+        {
+            var matcher = TryCreateSectionNameMatcher("[");
+            Assert.Null(matcher);
+        }
+
+        [Fact]
+        public void CharacterClassEscapedOpenEnded()
+        {
+            var matcher = TryCreateSectionNameMatcher(@"[\]");
+            Assert.Null(matcher);
+        }
+
+        [Fact]
+        public void CharacterClassEscapeAtEnd()
+        {
+            var matcher = TryCreateSectionNameMatcher(@"[\");
+            Assert.Null(matcher);
+        }
+
+        [Fact]
+        public void CharacterClassOpenBracketInside()
+        {
+            var matcher = TryCreateSectionNameMatcher(@"[[a]bc").Value;
+
+            Assert.True(matcher.IsMatch("/abc"));
+            Assert.True(matcher.IsMatch("/[bc"));
+            Assert.False(matcher.IsMatch("/ab"));
+            Assert.False(matcher.IsMatch("/[b"));
+            Assert.False(matcher.IsMatch("/bc"));
+            Assert.False(matcher.IsMatch("/ac"));
+            Assert.False(matcher.IsMatch("/[c"));
+
+            Assert.Equal(@"^.*/[\[a]bc$", matcher.Regex.ToString());
```

Comment:
Please put all the pattern assertions at the same point in the test. I don't really care whether it's first or last, but don't change it.